Rscript/Library/mzkit_app/man/mzkit/math/math/centroid.1
.\" man page create by R# package system.
.TH MATH 4 2000-1月 "centroid" "centroid"
.SH NAME
centroid \- Converts profiles peak data to peak data in centroid mode.
.SH SYNOPSIS
\fIcentroid(\fBions\fR as any, 
\fBtolerance\fR as any = "da:0.1", 
\fBintoCutoff\fR as double = 0.05, 
\fBparallel\fR as boolean = FALSE, 
\fBaggregate\fR as <generic> AggregateFunction = \fB<NULL>\fR, 
[\fB<Environment>\fR]);\fR
.SH DESCRIPTION
.PP
Converts profiles peak data to peak data in centroid mode.
 
 profile and centroid in Mass Spectrometry?
 
 1. Profile means the continuous wave form in a mass spectrum.
 
   + Number of data points Is large.
   
 2. Centroid means the peaks in a profile data Is changed to bars.
 
   + location of the bar Is center of the profile peak.
   + height of the bar Is area of the profile peak.
.PP
.SH OPTIONS
.PP
\fBaggregate\fB \fR\- 
 default is get the max intensity value.
. 
.PP
.PP
\fBions\fB \fR\- 
 value of this parameter could be 
 
 + a collection of peakMs2 data 
 + a library matrix data 
 + or a dataframe object which should contains at least \fBmz\fR and \fBintensity\fR columns.
 + or just a m/z vector
 + also could be a mzpack data object
 
. 
.PP
.SH VALUE
.PP
Peaks data in centroid mode or a new m/z vector in centroid.
.PP
.SH EXAMPLES
.PP
print(centroid([452.7627 67.563 457.336 347.8 242.3], tolerance = "da:0.1"));
 # [1]      67.563   242.3    347.8    452.763  457.336 
 
 let spec = data.frame(mz = [452.7627 67.563 457.336 347.8 242.3], 
    intensity = [312 4353 6664 6765 1119]);
    
 let f = aggregate(FUN = "max");
 
 print(as.data.frame(spec));
 #              mz intensity                   
 # --------------------------                                                                                                 
 # <mode> <Double> <integer>                                                                                                   
 # [1, ]   452.763       312                                             
 # [2, ]    67.563      4353                                                                    
 # [3, ]   457.336      6664                                                                                                
 # [4, ]     347.8      6765                                                                                           
 # [5, ]     242.3      1119
 
 print(centroid(spec, tolerance = "da:0.1", aggregate = f));
.PP
.SH SEE ALSO
math
.SH FILES
.PP
mzkit.dll
.PP
.SH AUTHOR
Written by \fBgg.xie@bionovogene.com\fR


HISTORY.md
# Update History

Mzkit is an open source raw data file toolkit for mass spectrometry data analysis, provides by the BioNovoGene corporation. The features of mzkit inlcudes: raw data file content viewer(XIC/TIC/Mass spectrum plot), build molecule network, formula de-novo search and de-novo annotation.

This open source mass spectrometry data toolkit is developed at the BioDeep R&D laboratory and brought to you by BioNovoGene corporation.

## ``3.5.593.3968`` Sep 1, 2022

> 10.5281/zenodo.7040586

+ ``new`` add tissue morphology map overlaps to MSI viewer
+ ``new`` add tissue morphology map editor feature to create custom tissue region for data analysis 
+ ``new`` add HE stain image analysis module
+ ``new`` add a new general table data viewer for open microsoft excel table files 
+ ``new`` add a general data visualization module for plot data based on the table viewer content
+ ``new`` add new ggplot package for data plot pipeline task
+ ``enhancement`` add data visualization template rendering for MS-imaging plot
+ ``enhancement`` enable view multiple sample MS-imaging data
+ ``enhancement`` add online pubchem metabolite database query function for the ion feature in MSI raw data 
+ ``enhancement`` add data compatibility with the bruker SCiLS lab software
+ ``enhancement`` make improvements of the ms1 peak list data annotation function 
+ ``enhancement`` update internal metabolite database, extends database list from KEGG only to KEGG/lipidmaps/chebi
+ ``enhancement`` make the molecular networking viewer interactive
+ ``enhancement`` add ms1 peak deconvolution function to raw data viewer
+ ``enhancement`` add peak finding analysis feature to the general signal data analysis 
+ ``enhancement`` new application installer experience
+ ``enhancement`` upgrade the internal Rstudio environment from .NET4.8 assembly to .NET6.0 assembly
+ ``enhancement`` add mzwork project file for share the workspace between the device
+ ``enhancement`` make improvements of the MRM/GCMS targetted data viewer
+ ``fixed`` add fix patch script to install internal Rstudio environment
+ ``fixed`` upgrade mzPack format to version 2.0, make improvements of the data compatibility between the single sample data and multiple sample data
+ ``fixed`` make bugs fixed of the style tweaking for raw data plot viewer
+ ``fixed`` handling of the data serialization error when raw data file its file path is a windows UNC path

## ``1.32.7743.6345`` Mar 14, 2021

> 10.5281/zenodo.4603277

+ ``enhancement`` improvements of the raw data charting plot styles
+ ``new`` add peak annotation result for MS matrix viewer
+ ``new`` targetted quantification linear modelling and sample quantify evaluation (required login BioDeep web services)
+ ``enhancement`` improvements of the GCMS/LC-MSMS targetted file explorer 
+ ``enhancement`` improvements of the GCMS feature ROI explorer
+ ``enhancement`` improvements of the document page model
+ ``fixed`` fixed of R# script editor syntax highlight problem
+ ``update`` update to latest mzkit R# package
+ ``new`` add 3D plot of LC-MSMS MRM ion TIC overlaps plot
+ ``enhancement`` add background task progress display on the status bar
+ ``new`` add demo scripts into the file explorer

## ``1.32.7692.29666`` Jan 22, 2021

> 10.5281/zenodo.4456618

+ ``enhancement`` Add plot tweaks function
+ ``new`` MS-imaging
+ ``new`` View MRM/GCMS SIM data files
+ ``enhancement`` Improvements of the R# scripting editor
+ ``enhancement`` Update R# interpreter engine

## ``v1.0.0.2-beta`` Nov 15, 2020

> 10.5281/zenodo.4274582

+ ``fixed`` Bugs fixed of the mzkit_win32 UI
+ ``enhancement`` Improvements of the content project file model
+ ``new`` Mass spectrum similarity search

## ``v1.0.0.1-beta`` Oct 15, 2020

> 10.5281/zenodo.4091067

+ ``new`` View mzML/mzXML raw data files
+ ``new`` Formula de-novol search
+ ``new`` Molecular networking
+ ``new`` Feature search
+ ``new`` R# scripting for mzkit automation


Rscript/R/mzkit/R/centroid.R
#Region "Microsoft.ROpen::46b8f7145b3fffcc63e247555d9b0b19, R\centroid.R"

    # Summaries:

    # centroid.2 <- function(profile, peakwidth = 0.1, intocutoff = 0.05) {if (!all(c("mz", "into") %in% colnames(profile))) {...

#End Region

#' Convert profile data to centroid
#'
#' @details Convert the profiles spectrum data matrix to centroid data mode.
#'    The matrix data for this function shoule be a dataframe object which
#'    at least contains 2 column data, where:
#'       \code{profile <- data.frame(mz = mzlist, into = intolist);}
#'
#'    If the \code{mz} or \code{into} data column is missing, then this function
#'    will throw an exception
#'
#' @param profile A 2D spectra data matrix in profile mode
#' @param peakwidth The spectra peak width in \code{da} unit
#'
#' @return A 2D spectra data matrix in simple centroid mode.
#'
centroid.2 <- function(profile, peakwidth = 0.1, intocutoff = 0.05) {
  if (!all(c("mz", "into") %in% colnames(profile))) {
    stop("Invalid prpfile spectra data matrix object!");
  }

  mz   = as.numeric(as.vector(profile[, "mz"]));
  into = as.numeric(as.vector(profile[, "into"]));

  # reorder in asc order
  i    = order(mz);
  mz   = mz[i];
  into = into[i];

  i    = (into / max(into)) >= intocutoff;
  mz   = mz[i];
  into = into[i];

  mzgroups = numeric.group(mz, assert = function(x, y) abs(x - y) <= peakwidth);
  cmz      = NULL;
  cinto    = NULL;

  for(mzi in mzgroups) {
    i     = (mz >= min(mzi)) & (mz <= max(mzi));
    int   = into[i];
    mzi   = mz[i];
    cmz   = append(cmz, mzi[which.max(int)]);
    cinto = append(cinto, max(int));
  }

  # we get a ms2 spectra peaks data in centroid mode
  data.frame(mz = cmz, into = cinto);
}

#' Convert \code{mzInto} object as data.frame
#' 
#' @param x a spectrum object in \code{mzInto} class.
#' 
#' @return a dataframe with two column fields: 
#'    \code{mz} and \code{into}.
#'
toMsMatrix = function(x) {
  data.frame(
    mz = x@mz,
    into = x@intensity
  );
}

#' Convert MS dataframe as \code{mzInto} class object.
#' 
#' @param x a spectrum data object in dataframe format. the first column
#'    in dataframe should be \code{m/z} value and the second column
#'    in dataframe should be \code{intensity} data.
#' 
toMzInto = function(x) {
  new("mzInto", 
    mz = as.numeric(as.vector(x[, 1])), 
    intensity = as.numeric(as.vector(x[, 2]))
  );
}